### DISCUSSION TITLE
Use triple slash directives for Auto-Import intellisense

### DISCUSSION THREAD

**Discussion by raphaelrk** [oven-sh/bun#15170] [Ideas]
It's possible to use triple slash directives for intellisense support on [auto-import](https://bun.sh/docs/runtime/autoimport) code

For example:

```tsx
#!/Users/myuser/.bun/bin/bun --hot
/// <reference types="/Users/myuser/.bun/install/cache/bun-types@1.1.34@@@1/index.d.ts" />
/// <reference types="/Users/myuser/.bun/install/cache/@types/node@20.12.14@@@1/index.d.ts" />
/// <reference types="/Users/myuser/.bun/install/cache/@types/ws@8.5.13@@@1/index.d.ts" />

const server = Bun.serve({
  fetch(req: Request) {
    return new Response("Success!");
  },
});

console.log(`Listening on http://localhost:${server.port}`);
```

The node and ws references are because those don't seem to get recognized by default — here's a screenshot of `bun-types@1.1.34@@@1/index.d.ts`:

<img width="617" alt="image" src="https://github.com/user-attachments/assets/9882541a-09f0-4e54-b0ed-0be3370c435a">

You can also move the triple-slash directives to a shared type declarations file, eg putting this in `~/bun-global-types/index.d.ts`

```tsx
/// <reference types="/Users/myuser/.bun/install/cache/bun-types@1.1.34@@@1/index.d.ts" />
/// <reference types="/Users/myuser/.bun/install/cache/@types/node@20.12.14@@@1/index.d.ts" />
/// <reference types="/Users/myuser/.bun/install/cache/@types/ws@8.5.13@@@1/index.d.ts" />
```

And then reference that in your code instead:

```tsx
/// <reference types="/Users/myuser/bun-global-types/index.d.ts" />
```

You can also omit the `index.d.ts`:

```tsx
/// <reference types="/Users/myuser/bun-global-types" />
```

The main issue I have with this approach is that it's a bit annoying finding/updating the paths for the type declarations. I wonder if it would be possible for bun to support a version of this that allows intellisense support without requiring me to look up the specific "packagename+version" paths, eg if I could just write: 

```tsx
/// <reference types="/Users/myuser/.bun/cache/bun" />
```

Or some variant of the above, some other possible examples:

```tsx
/// <reference types="/Users/myuser/.bun/auto-install-types/bun" />
/// <reference types="/Users/myuser/.bun/auto-install-types" />
/// <reference types="/Users/myuser/.bun/per-project-auto-install-types/my-project-name" />
/// <reference types="./.bun/auto-install-types" />
```

An IDE extension that adds support for auto-install types, fetching them from the bun global cache, could also be a nice solution.

**Comment by Jarred-Sumner** [comment]
interesting

**Reply by raphaelrk** [reply]
Thank you. I really like that bun makes "single-file-projects" possible. Better types support would be very nice, whether it uses what I wrote above or not, and whether it comes from the bun team or not
